cupy/cuda/stream.pyx

```diff
@@ -24,42 +40,40 @@ cdef class _ThreadLocal:
             tls = _thread_local.tls = _ThreadLocal()
         return <_ThreadLocal>tls
 
+    cdef void push_stream(self, stream, int device_id) except*:
+        assert device_id >= 0
+        self.current_stream_stack[device_id].append(stream)
+        # record device_id to prevent from popping the wrong stream at exit
+        self.current_device_id_stack.append(device_id)
+        self.set_current_stream(stream)
+
+    cdef void pop_stream(self) except*:
```

Comment:
How about passing a `device_id` from `__exit__`? I think `current_device_id_stack` is not needed in that case.

Reply:
Oh...This reminds me of a question: Do we want to handle the interaction between `Stream` and `Device` objects bidirectionally? The current behavior is:
```python
>>> cp.cuda.Device(0).use()
>>> s = cp.cuda.Stream()
>>> with cp.cuda.Device(1):
...     with s:
...         print(cp.cuda.get_current_stream())  # raise as we check if stream is valid on the device
... 
Traceback (most recent call last):
  File "<stdin>", line 2, in <module>
  File "cupy/cuda/stream.pyx", line 217, in cupy.cuda.stream.BaseStream.__enter__
    device_id = check_stream_device_match(device_id)
  File "cupy/cuda/stream.pyx", line 186, in cupy.cuda.stream.check_stream_device_match
    raise RuntimeError(
RuntimeError: This stream was not created on device 1
```
and
```python
>>> cp.cuda.Device(0).use()
>>> s = cp.cuda.Stream()
>>> with s:
...     with cp.cuda.Device(1):
...         print(cp.cuda.get_current_stream())  # we don't raise here <---- should we?
... 
<Stream 0 (device -1)>
```
If we remove `current_device_id_stack` and pass `device_id` from `__exit__`, we need to get the current device, which could be different from the entrant device (if `Device(1).use()` is called above).

Reply:
> If we remove current_device_id_stack and pass device_id from __exit__, we need to get the current device, which could be different from the entrant device (if Device(1).use() is called above).

Ah, ok, now I understood `current_device_id_stack` is needed for streams with device_id = -1 (singletons and ExtenalStream). Thank you for the explanation @leofang!

Reply:
> This reminds me of a question: Do we want to handle the interaction between `Stream` and `Device` objects bidirectionally?

Reply to self: No. See https://github.com/cupy/cupy/pull/5172#discussion_r625649470.